### PR TITLE
Fix bug when using distinct() and count()

### LINKED ISSUE #5958: realmQuery.distinct().count() might return incorrect value
A mapping of a Stack Overflow question that is most likely a bug.

#### Goal

> What do you want to achieve?

Evaluate number of distinct items without having to use `findAll().size()`

#### Expected Results

`realm.where(Blah.class).distinct("birthday").count()` should work

#### Actual Results

Count returns the size of the full table data set.

#### Steps & Code to Reproduce

See https://stackoverflow.com/questions/50363904/realm-giving-wrong-result-count-on-distinct

#### Version of Realm and tooling
Realm version(s): 5.1.0

Realm sync feature enabled: no

### PR DESCRIPTION
**Pull request by cmelchior**
Fixes #5958

### CODE REVIEW

File: realm/realm-library/src/androidTest/java/io/realm/RealmQueryTests.java [hunk 1]
```diff
@@ -2278,6 +2279,20 @@ public void count() {
         assertEquals(TEST_DATA_SIZE, realm.where(AllTypes.class).count());
     }
 
+    // Verify that count works together with using distinct.
```
Comment (nhachicha) [inline realm/realm-library/src/androidTest/java/io/realm/RealmQueryTests.java]: with->when?

File: realm/realm-library/src/main/java/io/realm/internal/TableQuery.java [hunk 2]
```diff
@@ -614,6 +614,13 @@ public long count(long start, long end, long limit) {
         return nativeCount(nativePtr, start, end, limit);
     }
 
+    /**
+     * Returns only the number of matching objects.
+     * This method is very fast compared to evaluating a query completely, but it does not
+     * goes around any logic implemented in Object Store and other parts of the API that works
+     * on query results. So it the primary use case for this method is testing.
```
Comment (nhachicha) [inline realm/realm-library/src/main/java/io/realm/internal/TableQuery.java]: So it the primary -> So the primary

File: realm/realm-library/src/main/java/io/realm/RealmQuery.java [hunk 3]
```diff
@@ -1749,8 +1749,12 @@ public Date maximumDate(String fieldName) {
      */
     public long count() {
         realm.checkIfValid();
-
-        return this.query.count();
+        // The fastest way of doing `count()` is going through `TableQuery.count()`. Unfortunately
+        // doing this does not correctly apply all side effects of queries (like subscriptions). Also
```
Comment (nhachicha) [inline realm/realm-library/src/main/java/io/realm/RealmQuery.java]: > ... does not correctly apply all side effects of queries (like subscriptions)

the new implementation doesn't create a subscription either?

Comment (cmelchior) [reply realm/realm-library/src/main/java/io/realm/RealmQuery.java]: You are right, but if we decided to change that in the future going directly to `TableQuery` would not work correctly.

**Comment by cmelchior** [comment]
> Isn't it a bug in Core if Query#count() doesn't return the correct size when used with distinct?

Not really. At the core level, `distinct` is not part of the query API, but is a method on `TableView`. Which is also why it is a lot easier to go through ObjectStore `Results` which will mask that difference.

**Pull request merged**